solidity/test/submitERC721LogicCall.ts

```diff
@@ -0,0 +1,225 @@
+import chai from "chai";
+import { ethers } from "hardhat";
+import { solidity } from "ethereum-waffle";
+import { GravityERC721 } from "../typechain/GravityERC721";
+import { deployContractsERC721 } from "../test-utils/deployERC721";
+import {
+  getSignerAddresses,
+  signHash,
+  examplePowers,
+  ZeroAddress
+} from "../test-utils/pure";
+
+chai.use(solidity);
+const { expect } = chai;
+
+
+async function runTest(opts: {
+  wrongCaller?: boolean;
+  wrongERC721Owner?: boolean;
+  ERC721NotExist?: boolean;
+  ERC721NotInContract?: boolean;
+}) {
+
+  // Prep and deploy contract
+  // ========================
+  const signers = await ethers.getSigners();
+  const gravityId = ethers.utils.formatBytes32String("foo");
+  // This is the power distribution on the Cosmos hub as of 7/14/2020
+  let powers = examplePowers();
+  let validators = signers.slice(0, powers.length);
+  const {
+    gravity,
+    gravityERC721,
+    fakeGravity,
+    testERC721,
+    testERC20,
+    checkpoint
+  } = await deployContractsERC721(gravityId, validators, powers);
+
+  const TestGravityERC721Contract = await ethers.getContractFactory("GravityERC721");
+  const ERC721LogicContract = (await TestGravityERC721Contract.deploy(gravity.address)) as GravityERC721;
+
+  // Transfer out to Cosmos, locking coins
+  // =====================================
+  await testERC20.functions.approve(gravity.address, 1000);
+  await gravity.functions.sendToCosmos(
+    testERC20.address,
+    ethers.utils.formatBytes32String("myCosmosAddress"),
+    1000
+  );
+
+  const numTxs = 10;
+  const txPayloads = new Array(numTxs);
+  const txAmounts = new Array(numTxs);
+  const tokenIds = new Array(numTxs);
+  const destinations = new Array(numTxs);
+
+  for (let i = 0; i < numTxs; i++) {
+    await testERC721.functions.approve(gravityERC721.address, 1+i);
+    if (!opts.ERC721NotInContract) {
+      await gravityERC721.functions["sendERC721ToCosmos(address,string,uint256)"](
+        testERC721.address,
+        ethers.utils.formatBytes32String("myCosmosAddress"),
+        1+i);
+    }
+    tokenIds[i] = 1+i;
+    destinations[i] = signers[i + 5].address;
+    txAmounts[i] = 0;
+  }
+
+  if (opts.wrongERC721Owner) {
+    const nftTokenOffset = 190; 
+    for (let i = nftTokenOffset; i < nftTokenOffset+numTxs; i++) {
```

Comment:
how does this offset translate to the wrong owner? Some comments or more descriptive logic would be great.

Reply:
Comment added